prototype/storage_api.py

```diff
@@ -0,0 +1,175 @@
+    # dictionary in its proper directory, I create a dictionary to associate the current path level with
+    # its param dict. After this loop, I create the dirs and then iterate through this dictionary to
+    # place the param dicts in their proper spots.
+    param_path_dict = {}
+    for view in mmif.views:
+        # this should return the back half of the app url, so just app name and version number
+        subdir_list = view.metadata.app.split('/')[3:]
+        # create path string for this view
+        view_path = os.path.join('', *subdir_list)
+        # now we want to convert the parameter dictionary to a string and then hash it.
+        # this hash will be the name of another subdirectory.
+        try:
+            param_dict = view.metadata["parameters"]
+            param_list = ['='.join(pair) for pair in param_dict.items()]
+            param_list.sort()
+            param_string = ','.join(param_list)
+        except KeyError:
+            param_dict = ""
+            param_string = ""
+        # hash the (sorted and concatenated list of params) string and join with path
+        # NOTE: this is *not* for security purposes, so the usage of md5 is not an issue.
+        param_hash = hashlib.md5(param_string.encode('utf-8')).hexdigest()
+        view_path = os.path.join(view_path, param_hash)
+        # check if this is a duplicate view. if it is, skip the current view.
+        # NOTE: duplicate views are those with the same app, version number, AND parameter dict.
+        if view_path in directory:
+            continue
+        # create path by joining directory with the current view path
+        directory = os.path.join(directory, view_path)
+        # now that we know it's not a duplicate view and we have the proper path location, we
+        # store it and the associated param dict inside param_path_dict.
+        param_path_dict[directory] = param_dict
+    # we have finished looping through the views. now time to create the directories
+    # and dump the param dicts
+    os.makedirs(directory, exist_ok=True)
+    for path in param_path_dict:
+        file_path = os.path.join(path, 'parameters.json')
+        with open(file_path, "w") as f:
+            json.dump(param_path_dict[path], f)
+    # put mmif into the lowest level directory with filename based on guid
+    file_path = os.path.join(directory, document)
+    with open(file_path, "w") as f:
+        f.write(mmif.serialize())
+    return "Success", 201
+
+
+@app.route("/retrieve/", methods=["POST"])
+def download_mmif():
+    # if not request.is_json:
+    #     return {'error': 'Request must be JSON'}, 400
+    data = json.loads(request.data.decode('utf-8'))
+    # get both pipeline and guid from data
+    # obtain pipeline using helper method
+    pipeline = pipeline_from_param_json(data)
+    # get number of views for rewind if necessary
+    num_views = len(data['pipeline'])
+    guid = data.get('guid')
```

Comment:
I think you mentioned you want to work on (or already on it?) multi-guid query scenario. As mentioned in other comment, how about also adding zero-guid query, to return just the full directory path?